File: buildy_lib/cache.py
"""Content-addressable build cache with thread-safe operations"""

import os
import json
import time
import hashlib
import shutil
import fcntl
import logging
from pathlib import Path
from typing import Dict, List, Any

from .models import BuildTask

logger = logging.getLogger('buildy.cache')

class BuildCache:
    """Content-addressable build cache with thread-safe operations"""

    def __init__(self, cache_dir: str = ".buildy_cache"):
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(exist_ok=True)
        self.cache_index_file = self.cache_dir / "cache_index.json"
        self.cache_lock_file = self.cache_dir / "cache.lock"
        self.objects_dir = self.cache_dir / "objects"  # Git-style objects directory
        self.objects_dir.mkdir(exist_ok=True)
        self.cache_index = self._load_cache_index()

    def _get_cache_path(self, cache_key: str) -> Path:
        """Get sharded cache directory path using Git-style sharding
        
        Uses first 2 hex characters as subdirectory to distribute files
        across 256 directories (00-ff) for better filesystem performance.
        
        Example:
            cache_key = '6ec1a1ed4e3693526c4ce9e5b89215f860c6a853eacf1f9b3fe5bfc10c90b580'
            returns: .buildy_cache/objects/6e/c1a1ed4e3693526c4ce9e5b89215f860c6a853eacf1f9b3fe5bfc10c90b580/
        
        This scales to millions of cached tasks without filesystem performance degradation.
        """
        shard = cache_key[:2]  # First 2 hex characters (00-ff)
        return self.objects_dir / shard / cache_key
    
    def _load_cache_index(self) -> Dict[str, Any]:
        """Load cache index from disk with locking"""
        if self.cache_index_file.exists():
            try:
                with open(self.cache_index_file, 'r') as f:
                    # Acquire shared lock for reading
                    fcntl.flock(f.fileno(), fcntl.LOCK_SH)
                    try:
                        data = json.load(f)
                        return data
                    finally:
                        fcntl.flock(f.fileno(), fcntl.LOCK_UN)
            except (json.JSONDecodeError, IOError) as e:
                logger.warning(f"Failed to load cache index: {e}")
        return {}

    def _save_cache_index(self):
        """Save cache index to disk atomically with locking"""
        try:
            # Ensure cache directory exists (may be deleted in parallel builds)
            self.cache_dir.mkdir(parents=True, exist_ok=True)
            
            # Write to temporary file first
            temp_file = self.cache_index_file.with_suffix('.tmp')
            with open(temp_file, 'w') as f:
                # Acquire exclusive lock for writing
                fcntl.flock(f.fileno(), fcntl.LOCK_EX)
                try:
                    json.dump(self.cache_index, f, indent=2)
                    f.flush()
                    os.fsync(f.fileno())
                finally:
                    fcntl.flock(f.fileno(), fcntl.LOCK_UN)
            
            # Atomic rename
            temp_file.replace(self.cache_index_file)
        except (OSError, IOError) as e:
            logger.error(f"Failed to save cache index: {e}")

    def has_cached_result(self, task: BuildTask) -> bool:
        """Check if task result is cached and all dependencies are unchanged"""
        cache_entry = self.cache_index.get(task.cache_key)
        if not cache_entry:
            return False

        # Check if all output files exist and match cached hashes
        for output_path, expected_hash in cache_entry.get('outputs', {}).items():
            if not os.path.exists(output_path):
                logger.debug(f"Cache miss for {task.task_id}: output {output_path} not found")
                return False

            # Directories are always considered valid if they exist
            if expected_hash == "directory":
                continue
            
            actual_hash = self._calculate_file_hash(output_path)
            if actual_hash != expected_hash:
                logger.debug(f"Cache miss for {task.task_id}: output {output_path} hash changed")
                return False
        
        # Check header dependencies (for compile tasks)
        header_hashes = cache_entry.get('header_hashes', {})
        if header_hashes:
            for header_path, expected_hash in header_hashes.items():
                # Check if header still exists
                if not os.path.exists(header_path):
                    logger.debug(f"Cache miss for {task.task_id}: header {header_path} deleted")
                    return False
                
                # Check if header content changed
                actual_hash = self._calculate_file_hash(header_path)
                if actual_hash != expected_hash:
                    logger.info(f"Cache miss for {task.task_id}: header {header_path} modified")
                    return False
            
            logger.debug(f"Cache hit for {task.task_id}: {len(header_hashes)} headers unchanged")

        return True

    def restore_cached_result(self, task: BuildTask) -> bool:
        """Restore cached task outputs"""
        cache_entry = self.cache_index.get(task.cache_key)
        if not cache_entry:
            return False

        # Use sharded path for better filesystem performance
        cache_files_dir = self._get_cache_path(task.cache_key)
        if not cache_files_dir.exists():
            return False

        try:
            # Restore output files from cache (skip directories)
            for output_path in task.outputs:
                # Skip directories - they should be created by the task if needed
                if output_path.endswith('/'):
                    os.makedirs(output_path, exist_ok=True)
                    continue
                    
                cached_file = cache_files_dir / Path(output_path).name
                if cached_file.exists():
                    output_dir = os.path.dirname(output_path)
                    if output_dir:
                        os.makedirs(output_dir, exist_ok=True)
                    shutil.copy2(cached_file, output_path)

            logger.info(f"✓ {task.task_id} - cache hit, restored outputs")
            return True
        except PermissionError as e:
            logger.error(f"✗ {task.task_id} - permission denied: {e}")
            return False
        except OSError as e:
            logger.error(f"✗ {task.task_id} - cache restore failed: {e}")
            return False
        except Exception as e:
            logger.error(f"✗ {task.task_id} - unexpected error during restore: {e}")
            return False

    def cache_task_result(self, task: BuildTask, execution_time: float, success: bool):
        """Cache task result after successful execution"""
        if not success:
            return

        try:
            # Use sharded path for better filesystem performance
            cache_files_dir = self._get_cache_path(task.cache_key)
            cache_files_dir.mkdir(parents=True, exist_ok=True)

            # Cache output files (skip directories)
            output_hashes = {}
            dep_file = None
            for output_path in task.outputs:
                if os.path.exists(output_path):
                    # Skip directories - they can't be cached as files
                    if os.path.isdir(output_path):
                        output_hashes[output_path] = "directory"
                        continue
                    
                    # Track .d file for header dependency parsing
                    if output_path.endswith('.d'):
                        dep_file = output_path
                    
                    cached_file = cache_files_dir / Path(output_path).name
                    shutil.copy2(output_path, cached_file)
                    output_hashes[output_path] = self._calculate_file_hash(output_path)
            
            # Parse header dependencies from .d file if this is a compile task
            header_deps = []
            header_hashes = {}
            if dep_file and task.task_type == 'compile':
                header_deps = self._parse_dependency_file(dep_file)
                # Calculate and store hashes for all header dependencies
                for header_path in header_deps:
                    if os.path.exists(header_path):
                        header_hashes[header_path] = self._calculate_file_hash(header_path)
                    else:
                        logger.debug(f"Header dependency not found: {header_path}")

            # Update cache index
            self.cache_index[task.cache_key] = {
                'task_id': task.task_id,
                'task_type': task.task_type,
                'cached_at': time.time(),
                'execution_time': execution_time,
                'outputs': output_hashes,
                'header_dependencies': header_deps,  # NEW: List of header paths
                'header_hashes': header_hashes,      # NEW: Header path -> hash mapping
                'platform': task.platform,
                'architecture': task.architecture,
                'configuration': task.configuration
            }
            self._save_cache_index()

        except PermissionError as e:
            logger.error(f"✗ {task.task_id} - permission denied during caching: {e}")
        except OSError as e:
            logger.error(f"✗ {task.task_id} - caching failed: {e}")
        except Exception as e:
            logger.error(f"✗ {task.task_id} - unexpected error during caching: {e}")

    def _calculate_file_hash(self, file_path: str) -> str:
        """Calculate SHA-256 hash of file efficiently using chunked reading"""
        hash_obj = hashlib.sha256()
        try:
            with open(file_path, 'rb') as f:
                # Read in 8KB chunks for memory efficiency with large files
                for chunk in iter(lambda: f.read(8192), b''):
                    hash_obj.update(chunk)
            return hash_obj.hexdigest()
        except (OSError, IOError) as e:
            logger.warning(f"Failed to hash file {file_path}: {e}")
            return ""
    
    def _parse_dependency_file(self, dep_file: str) -> List[str]:
        """Parse GCC-generated .d file to extract header dependencies
        
        .d file format example:
        build/obj/math.o: examples/math/libsrc/math.cpp \\
          examples/math/include/mymath.h \\
          /usr/include/c++/11/iostream
        
        Returns list of header file paths (relative or absolute)
        """
        if not os.path.exists(dep_file):
            return []
        
        try:
            with open(dep_file, 'r', encoding='utf-8') as f:
                content = f.read()
            
            # Remove target (everything before and including ':')
            if ':' not in content:
                return []
            
            content = content.split(':', 1)[1]
            
            # Remove line continuations (backslash + newline)
            content = content.replace('\\\n', ' ').replace('\\', '')
            
            # Split on whitespace and filter
            all_deps = content.split()
            
            # Filter to only header files (skip .cpp, .c source files)
            headers = [
                dep.strip() for dep in all_deps 
                if dep.strip() and dep.endswith(('.h', '.hpp', '.hxx', '.hh', '.H'))
            ]
            
            logger.debug(f"Parsed {len(headers)} header dependencies from {dep_file}")
            return headers
            
        except Exception as e:
            logger.warning(f"Failed to parse dependency file {dep_file}: {e}")
            return []

    def get_cache_stats(self) -> Dict[str, Any]:
        """Get cache statistics"""
        total_entries = len(self.cache_index)
        total_size = 0

        # Iterate through sharded cache directories
        for cache_key in self.cache_index:
            cache_dir = self._get_cache_path(cache_key)
            if cache_dir.exists():
                for file_path in cache_dir.rglob('*'):
                    if file_path.is_file():
                        total_size += file_path.stat().st_size

        return {
            'total_entries': total_entries,
            'total_size_mb': total_size / (1024 * 1024),
            'cache_directory': str(self.cache_dir)
        }

File: src/internal/cache/cache.go
package cache

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"buildy/internal/workspace"
)

// CacheEntry represents a cached build task result
type CacheEntry struct {
	TaskID             string            `json:"task_id"`
	TaskType           string            `json:"task_type"`
	CachedAt           float64           `json:"cached_at"`
	ExecutionTime      float64           `json:"execution_time"`
	Outputs            map[string]string `json:"outputs"`             // path -> hash
	HeaderDependencies []string          `json:"header_dependencies"` // list of header paths
	HeaderHashes       map[string]string `json:"header_hashes"`       // path -> hash
	Platform           string            `json:"platform"`
	Architecture       string            `json:"architecture"`
	Configuration      string            `json:"configuration"`
}

// BuildCache provides content-addressable build cache with thread-safe operations
type BuildCache struct {
	cacheDir       string
	cacheIndexFile string
	cacheLockFile  string
	objectsDir     string
	cacheIndex     map[string]CacheEntry
	mutex          sync.RWMutex // For thread-safe cache index access
}

// NewBuildCache creates a new build cache
func NewBuildCache(cacheDir string) (*BuildCache, error) {
	if cacheDir == "" {
		cacheDir = ".buildy_cache"
	}

	cache := &BuildCache{
		cacheDir:       cacheDir,
		cacheIndexFile: filepath.Join(cacheDir, "cache_index.json"),
		cacheLockFile:  filepath.Join(cacheDir, "cache.lock"),
		objectsDir:     filepath.Join(cacheDir, "objects"),
		cacheIndex:     make(map[string]CacheEntry),
	}

	// Create cache directory structure
	if err := os.MkdirAll(cache.cacheDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create cache directory: %w", err)
	}

	if err := os.MkdirAll(cache.objectsDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create objects directory: %w", err)
	}

	// Load cache index
	if err := cache.loadCacheIndex(); err != nil {
		log.Printf("Warning: failed to load cache index: %v", err)
	}

	return cache, nil
}

// CacheDir returns the cache directory path
func (bc *BuildCache) CacheDir() string {
	return bc.cacheDir
}

// getCachePath returns sharded cache directory path using Git-style sharding
// Uses first 2 hex characters as subdirectory to distribute files
// across 256 directories (00-ff) for better filesystem performance.
func (bc *BuildCache) getCachePath(cacheKey string) string {
	if len(cacheKey) < 2 {
		return filepath.Join(bc.objectsDir, cacheKey)
	}
	shard := cacheKey[:2] // First 2 hex characters (00-ff)
	return filepath.Join(bc.objectsDir, shard, cacheKey)
}

// sanitizeCacheRelPath produces a cache-internal relative path for a build output.
// We preserve the original directory structure to avoid filename collisions, but
// strip drive letters and leading separators so the path remains relative.
func sanitizeCacheRelPath(outputPath string) string {
	cleaned := filepath.Clean(outputPath)

	// Drop drive letters (Windows) to avoid creating top-level drive dirs
	cleaned = strings.ReplaceAll(cleaned, ":", "")

	// Make absolute paths relative to root so they can live inside cache dir
	if filepath.IsAbs(cleaned) {
		cleaned = strings.TrimPrefix(cleaned, string(filepath.Separator))
	}

	return cleaned
}

// loadCacheIndex loads cache index from disk
func (bc *BuildCache) loadCacheIndex() error {
	bc.mutex.Lock()
	defer bc.mutex.Unlock()

	if _, err := os.Stat(bc.cacheIndexFile); os.IsNotExist(err) {
		return nil // No cache index yet, that's fine
	}

	data, err := os.ReadFile(bc.cacheIndexFile)
	if err != nil {
		return fmt.Errorf("failed to read cache index: %w", err)
	}

	if err := json.Unmarshal(data, &bc.cacheIndex); err != nil {
		return fmt.Errorf("failed to parse cache index: %w", err)
	}

	return nil
}

// saveCacheIndex saves cache index to disk atomically (acquires lock)
func (bc *BuildCache) saveCacheIndex() error {
	bc.mutex.RLock()
	defer bc.mutex.RUnlock()
	return bc.saveCacheIndexLocked()
}

// saveCacheIndexLocked saves cache index to disk atomically
// IMPORTANT: Caller must hold at least a read lock on bc.mutex
func (bc *BuildCache) saveCacheIndexLocked() error {
	// Ensure cache directory exists
	if err := os.MkdirAll(bc.cacheDir, 0755); err != nil {
		return fmt.Errorf("failed to create cache directory: %w", err)
	}

	// Create temp file
	tempFile, err := os.CreateTemp(bc.cacheDir, "cache_index_*.tmp")
	if err != nil {
		return fmt.Errorf("failed to create temp file: %w", err)
	}
	tempPath := tempFile.Name()

	// Write JSON
	encoder := json.NewEncoder(tempFile)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(bc.cacheIndex); err != nil {
		tempFile.Close()
		os.Remove(tempPath)
		return fmt.Errorf("failed to encode cache index: %w", err)
	}

	// Sync and close
	if err := tempFile.Sync(); err != nil {
		tempFile.Close()
		os.Remove(tempPath)
		return fmt.Errorf("failed to sync temp file: %w", err)
	}

	if err := tempFile.Close(); err != nil {
		os.Remove(tempPath)
		return fmt.Errorf("failed to close temp file: %w", err)
	}

	// Atomic rename
	if err := os.Rename(tempPath, bc.cacheIndexFile); err != nil {
		os.Remove(tempPath)
		return fmt.Errorf("failed to rename temp file: %w", err)
	}

	return nil
}

// HasCacheEntry returns true if a cache entry exists for the given key
func (bc *BuildCache) HasCacheEntry(cacheKey string) bool {
	bc.mutex.RLock()
	defer bc.mutex.RUnlock()
	_, exists := bc.cacheIndex[cacheKey]
	return exists
}

// RestoreCachedResult restores cached task outputs
func (bc *BuildCache) RestoreCachedResult(task *workspace.BuildTask) error {
	bc.mutex.RLock()
	_, exists := bc.cacheIndex[task.CacheKey]
	bc.mutex.RUnlock()

	if !exists {
		return fmt.Errorf("cache entry not found")
	}

	cacheFilesDir := bc.getCachePath(task.CacheKey)
	if _, err := os.Stat(cacheFilesDir); os.IsNotExist(err) {
		return fmt.Errorf("cache directory not found")
	}

	// Restore output files from cache (skip directories)
	for _, outputPath := range task.Outputs {
		// Skip directories - they should be created by the task if needed
		if strings.HasSuffix(outputPath, string(filepath.Separator)) {
			if err := os.MkdirAll(outputPath, 0755); err != nil {
				return fmt.Errorf("failed to create directory %s: %w", outputPath, err)
			}
			continue
		}

		relPath := sanitizeCacheRelPath(outputPath)
		cachedFile := filepath.Join(cacheFilesDir, relPath)
		if _, err := os.Stat(cachedFile); err == nil {
			outputDir := filepath.Dir(outputPath)
			if outputDir != "" && outputDir != "." {
				if err := os.MkdirAll(outputDir, 0755); err != nil {
					return fmt.Errorf("failed to create output directory: %w", err)
				}
			}

			if err := copyFile(cachedFile, outputPath); err != nil {
				return fmt.Errorf("failed to copy cached file: %w", err)
			}
		}
	}

	log.Printf("✓ %s - cache hit, restored outputs", task.TaskID)
	return nil
}

// CacheTaskResult caches task result after successful execution
func (bc *BuildCache) CacheTaskResult(task *workspace.BuildTask, executionTime float64, success bool) error {
	if !success {
		return nil
	}

	cacheFilesDir := bc.getCachePath(task.CacheKey)
	if err := os.MkdirAll(cacheFilesDir, 0755); err != nil {
		return fmt.Errorf("failed to create cache directory: %w", err)
	}

	// Cache output files (skip directories)
	outputHashes := make(map[string]string)
	var depFile string

	for _, outputPath := range task.Outputs {
		if _, err := os.Stat(outputPath); err == nil {
			// Skip directories - they can't be cached as files
			info, err := os.Stat(outputPath)
			if err == nil && info.IsDir() {
				outputHashes[outputPath] = "directory"
				continue
			}

			// Track .d file for header dependency parsing
			if strings.HasSuffix(outputPath, ".d") {
				depFile = outputPath
			}

			relPath := sanitizeCacheRelPath(outputPath)
			cachedFile := filepath.Join(cacheFilesDir, relPath)
			if err := os.MkdirAll(filepath.Dir(cachedFile), 0755); err != nil {
				return fmt.Errorf("failed to create cached output directory: %w", err)
			}

			if err := copyFile(outputPath, cachedFile); err != nil {
				return fmt.Errorf("failed to cache file: %w", err)
			}

			hash, err := bc.calculateFileHash(outputPath)
			if err != nil {
				return fmt.Errorf("failed to hash output: %w", err)
			}
			outputHashes[outputPath] = hash
		}
	}

	// Parse header dependencies from .d file if this is a compile task
	var headerDeps []string
	headerHashes := make(map[string]string)

	if depFile != "" && task.TaskType == "compile" {
		headerDeps = bc.parseDependencyFile(depFile)
		// Calculate and store hashes for all header dependencies
		for _, headerPath := range headerDeps {
			if _, err := os.Stat(headerPath); err == nil {
				hash, err := bc.calculateFileHash(headerPath)
				if err == nil {
					headerHashes[headerPath] = hash
				}
			} else {
				log.Printf("Header dependency not found: %s", headerPath)
			}
		}
	}

	// Update cache index and save atomically while holding the lock
	bc.mutex.Lock()
	bc.cacheIndex[task.CacheKey] = CacheEntry{
		TaskID:             task.TaskID,
		TaskType:           task.TaskType,
		CachedAt:           float64(time.Now().Unix()),
		ExecutionTime:      executionTime,
		Outputs:            outputHashes,
		HeaderDependencies: headerDeps,
		HeaderHashes:       headerHashes,
		Platform:           task.Platform,
		Architecture:       task.Architecture,
		Configuration:      task.Configuration,
	}
	err := bc.saveCacheIndexLocked()
	bc.mutex.Unlock()

	return err
}

// calculateFileHash calculates SHA-256 hash of file efficiently using chunked reading
func (bc *BuildCache) calculateFileHash(filePath string) (string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer file.Close()

	hash := sha256.New()

	// Read in 8KB chunks for memory efficiency with large files
	buf := make([]byte, 8192)
	for {
		n, err := file.Read(buf)
		if n > 0 {
			hash.Write(buf[:n])
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			return "", err
		}
	}

	return fmt.Sprintf("%x", hash.Sum(nil)), nil
}

// parseDependencyFile parses compiler-generated dependency file
func (bc *BuildCache) parseDependencyFile(depFile string) []string {
	if _, err := os.Stat(depFile); os.IsNotExist(err) {
		return nil
	}

	// Detect format by extension
	if strings.HasSuffix(depFile, ".json") {
		return bc.parseMSVCJSONDeps(depFile)
	}
	return bc.parseMakefileDeps(depFile)
}

// parseMakefileDeps parses GCC/Clang Makefile-style .d file
func (bc *BuildCache) parseMakefileDeps(depFile string) []string {
	data, err := os.ReadFile(depFile)
	if err != nil {
		log.Printf("Warning: failed to read dependency file %s: %v", depFile, err)
		return nil
	}

	content := string(data)

	// Remove target (everything before and including ':')
	idx := strings.Index(content, ":")
	if idx == -1 {
		return nil
	}
	content = content[idx+1:]

	// Remove line continuations (backslash + newline)
	content = strings.ReplaceAll(content, "\\\n", " ")
	content = strings.ReplaceAll(content, "\\", "")

	// Split on whitespace
	allDeps := strings.Fields(content)

	// Filter to only header files (skip .cpp, .c source files)
	// Include .inl (inline implementation files) and .inc (include files)
	var headers []string
	for _, dep := range allDeps {
		dep = strings.TrimSpace(dep)
		if dep != "" && (strings.HasSuffix(dep, ".h") ||
			strings.HasSuffix(dep, ".hpp") ||
			strings.HasSuffix(dep, ".hxx") ||
			strings.HasSuffix(dep, ".hh") ||
			strings.HasSuffix(dep, ".H") ||
			strings.HasSuffix(dep, ".inl") ||
			strings.HasSuffix(dep, ".inc")) {
			headers = append(headers, dep)
		}
	}

	log.Printf("Parsed %d header dependencies from %s (Makefile format)", len(headers), depFile)
	return headers
}

// parseMSVCJSONDeps parses MSVC /sourceDependencies JSON format
func (bc *BuildCache) parseMSVCJSONDeps(depFile string) []string {
	data, err := os.ReadFile(depFile)
	if err != nil {
		log.Printf("Warning: failed to read dependency file %s: %v", depFile, err)
		return nil
	}

	var jsonData struct {
		Version string `json:"Version"`
		Data    struct {
			Source   string   `json:"Source"`
			Includes []string `json:"Includes"`
		} `json:"Data"`
	}

	if err := json.Unmarshal(data, &jsonData); err != nil {
		log.Printf("Warning: failed to parse JSON dependency file %s: %v", depFile, err)
		return nil
	}

	// Filter to only user headers (skip system headers in common system paths)
	var headers []string
	for _, inc := range jsonData.Data.Includes {
		// Normalize path separators
		incNormalized := strings.ReplaceAll(inc, "\\", "/")
		incLower := strings.ToLower(incNormalized)

		// Skip system headers in common locations
		if strings.HasPrefix(incLower, "c:/program files") ||
			strings.HasPrefix(incLower, "c:/windows") ||
			strings.HasPrefix(incLower, "/usr/include") ||
			strings.HasPrefix(incLower, "/usr/local/include") {
			continue
		}

		// Only include files with header extensions
		if strings.HasSuffix(incNormalized, ".h") ||
			strings.HasSuffix(incNormalized, ".hpp") ||
			strings.HasSuffix(incNormalized, ".hxx") ||
			strings.HasSuffix(incNormalized, ".hh") ||
			strings.HasSuffix(incNormalized, ".H") ||
			strings.HasSuffix(incNormalized, ".inl") ||
			strings.HasSuffix(incNormalized, ".inc") {
			headers = append(headers, filepath.FromSlash(incNormalized))
		}
	}

	log.Printf("Parsed %d header dependencies from %s (MSVC JSON format)", len(headers), depFile)
	return headers
}

// GetCacheStats returns cache statistics
func (bc *BuildCache) GetCacheStats() map[string]interface{} {
	bc.mutex.RLock()
	totalEntries := len(bc.cacheIndex)
	bc.mutex.RUnlock()

	var totalSize int64

	// Iterate through sharded cache directories
	filepath.Walk(bc.objectsDir, func(path string, info os.FileInfo, err error) error {
		if err == nil && !info.IsDir() {
			totalSize += info.Size()
		}
		return nil
	})

	return map[string]interface{}{
		"total_entries":   totalEntries,
		"total_size_mb":   float64(totalSize) / (1024 * 1024),
		"cache_directory": bc.cacheDir,
	}
}

// copyFile copies a file from src to dst
func copyFile(src, dst string) error {
	sourceFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer sourceFile.Close()

	destFile, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer destFile.Close()

	if _, err := io.Copy(destFile, sourceFile); err != nil {
		return err
	}

	// Copy file permissions
	sourceInfo, err := os.Stat(src)
	if err != nil {
		return err
	}

	return os.Chmod(dst, sourceInfo.Mode())
}
